### ISSUE TITLE
Params not yet available in 'pushState' and 'render' event handlers 

### ISSUE THREAD

**Issue by brechtpm** [choojs/choo#463]
### Expected behavior

When handling a 'pushState' event, the resolved `state.params` should already be available upon calling the event handler.

### Actual behavior

Resolved params are only added to state during render.

### Steps to reproduce behavior

To reproduce, click on the link to 'some page' in this example app:

```js
var choo = require('choo')
var html = require('choo/html')

var app = choo()

app.route('/', function () {
  return html`
    <body>
      <a href="/somepage">some page</a>
    </body>
  `
})

app.route('/*', function (state) {
  return html`
    <body>
      page: ${state.params.wildcard} <!-- correctly renders 'somepage' -->
    </body>
  `
})

app.use(function (state, bus) {
  bus.on('pushState', function () {
    console.log(state.params.wildcard) // logs undefined
  })

  bus.on('render', function () {
    console.log(state.params.wildcard) // logs undefined
  })
})

app.mount('body')
```

This is a problem for my specific use case, because in the 'pushState' (or 'render') event handler, I want to fetch some data based on the page param, and fill out the page based on the response.

As an aside, I do want to mention that working with `choo v5` has been a real joy so far. You absolutely nailed it with the new stripped-down API for `app.use`!

**Comment by brechtpm** [comment]
Hm, seems to be related to https://github.com/yoshuawuyts/choo/issues/447. I could close this one if you feel they're too similar.

**Comment by brechtpm** [comment]
Okay, so I've read through https://github.com/yoshuawuyts/choo/issues/447, and on second thought I think this issue is distinct enough to warrant its own thread. Feel free to close if you prefer to keep them together though.

**Comment by yoshuawuyts** [comment]
The reason this happens is because your `pushState` listener fires before the router gets a chance to pick it up. I agree t

I think the solution would be to implement [events.prependListener](https://nodejs.org/api/events.html#events_emitter_prependlistener_eventname_listener) (and the "once" equivalent) in [nanobus](https://github.com/yoshuawuyts/nanobus/), and then use that method instead of `.on()` to register the listener in `app.start()` - that way the router will always fire before other listeners which would solve your issue.

In the short term you can move your `.use()` calls to after `.start()` which should have the same effect.

**Comment by yoshuawuyts** [comment]
https://github.com/yoshuawuyts/nanobus/pull/9

**Comment by espy** [comment]
I’m afraid this is still a problem in the current choo (`6.6.0`), I’ve taken the app generated by `npx create-choo-app` and added the following:

to `store.js`
```javascript
  emitter.on('navigate', () => {
    console.log('navigated, state.params is', state.params)
  })
```
`index.js` gets one more route:
```javascript
app.route('/day/:date', require('./views/main'))
```
and `main.js`'s html output gets some links so we can get there.
```javascript
<a href='/day/2017-12-29'>2017-12-29</a>
<a href='/day/2017-12-30'>2017-12-30</a>
```
The resulting behaviour is:
first click: params is empty
second click: params.date is from the route we just left

So maybe this isn’t closed after all 😿 

As for workarounds, you wrote:
> In the short term you can move your .use() calls to after .start() which should have the same effect.

But there no longer seems to be a `.start()` anywhere, and moving the `.use()` to the end of `index.js` makes no difference. My workaround is currently a `setTimeout` 🙈

**Comment by yoshuawuyts** [comment]
Reopened. Yeah, this doesn't sound great - I'm sorry :(

**Comment by yoshuawuyts** [comment]
Should be fixed in `choo@6.7.0` :tada:
